userspace/libsinsp/filter/ast.cpp

```diff
@@ -16,9 +16,143 @@ limitations under the License.
 */
 
 #include "ast.h"
+#include "sinsp_exception.h"
 
 using namespace libsinsp::filter::ast;
 
+std::string libsinsp::filter::ast::unescape_str(const std::string& str)
+{
+	std::string res = "";
+	size_t len = str.size() - 1;
+	bool escaped = false;
+	for (size_t i = 1; i < len; i++)
+	{
+		if (!escaped)
+		{
+			if (str[i] == '\\')
+			{
+				escaped = true;
+			}
+			else
+			{
+				res += str[i];
+			}
+		}
+		else
+		{
+			switch(str[i])
+			{
+				case 'b':
+					res += '\b';
+					break;
+				case 'f':
+					res += '\f';
+					break;
+				case 'n':
+					res += '\n';
+					break;
+				case 'r':
+					res += '\r';
+					break;
+				case 't':
+					res += '\t';
+					break;
+				case ' ':
+					// NOTE: we may need to initially support this to not create breaking changes with
+					// some existing wrongly-escaped rules. So far, I only found one, in Falco:
+					// https://github.com/falcosecurity/falco/blob/204f9ff875be035e620ca1affdf374dd1c610a98/rules/falco_rules.yaml#L3046
+					// todo(jasondellaluce): remove this once rules are rewritten with correct escaping
+				case '\\':
+					res += '\\';
+					break;
+				case '/':
+					res += '/';
+					break;
+				case '"':
+					if (str[0] != str[i])
+					{
+						throw sinsp_exception("invalid \\\" escape in '-quoted string");
+					}
+					res += '\"';
+					break;
+				case '\'':
+					if (str[0] != str[i])
+					{
+						throw sinsp_exception("invalid \\' escape in \"-quoted string");
+					}
+					res += '\'';
+					break;
+				case 'x':
+					// todo(jasondellaluce): support hex num escaping (not needed for now)
+				default:
+					throw sinsp_exception("unsupported string escape sequence: \\" + std::string(1, str[i]));
+			}
+			escaped = false;
+		}
+	}
+	return res;
+}
+
+std::string libsinsp::filter::ast::escape_str(const std::string& str)
+{
+	std::string res = "";
+	char escape_char = '"';
+	size_t len = str.size();
+	bool should_escape = false;
+	for (size_t i = 0; i < len; i++)
+	{
+		switch(str[i])
+		{
+		case '\b':
+			should_escape = true;
+			res += "\\b";
+			break;
+		case '\f':
+			should_escape = true;
+			res += "\\f";
+			break;
+		case '\n':
+			should_escape = true;
+			res += "\\n";
+			break;
+		case '\r':
+			should_escape = true;
+			res += "\\r";
+			break;
+		case '\t':
+			should_escape = true;
+			res += "\\t";
+			break;
+		case ' ':
+			should_escape = true;
+			res += ' ';
+			break;
+		case '\\':
+			should_escape = true;
+			res += "\\\\";
+			break;
+		case '"':
+			should_escape = true;
+			res += "\\\"";
+			break;
+		case '\'':
+			should_escape = true;
+			escape_char = '\'';
```

Comment:
Note, this will not work. There may be a case in which you already escaped a `"` as `\"` before reaching this point, and this would mix `'` escape delimiters with `\"`. Now, I see two solutions to this:
1. **More conservative**: we decide to always have `"` as an escape delimiter and we remove this line 140. In this way, strings are always escaped with `"` which can result in uglier results in some edge case but it's never ambigue.
2. **Less conservative**: we drop the assumption that delimiters `"` and `'` can't be respectively mixed with `\'` and `\"` (enforced at line 74 and 81 here in `unescape_str()`). This would allow you to achieve what you wanted, but has the downside of potentially creating breaking changes in many existing rulesets, and also makes the grammar more loose and error-prone.

Due to the benefits/downsides tradeoffs described above, my vote goes for option 1). WDYT?

Reply:
I'll do the first thing--always escape with double quotes. That will mean that an expression like `proc.name in ('some proc')` will be parsed and printed as `proc.name in (\"some proc\")`. This breaks the assumption that parsed-and-printed result in the same expression, but if parsed again, it *will* result in an equal ast.

Reply:
I think having equal ASTs is the bigger concern here.